src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts

```diff
@@ -107,6 +107,37 @@ const isExternalLink = (url: string): boolean => {
     return !url.startsWith(window.location.origin);
 };
 
+const toggleBlockType = (schema, type, attrs = {}, wrap = false) => {
+    const blockType = schema.nodes[type];
+    const paragraphType = schema.nodes.paragraph;
+
+    return (state, dispatch) => {
+        const { $from, to } = state.selection;
+        if (
+            state.selection instanceof TextSelection &&
+            $from.sameParent($from.doc.resolve(to))
+        ) {
+            if ($from.parent.type === blockType) {
+                if (dispatch) {
+                    dispatch(
+                        state.tr.setBlockType($from.pos, to, paragraphType),
+                    );
+                }
+
+                return true;
+            } else {
+                if (wrap) {
+                    return wrapIn(blockType, attrs)(state, dispatch);
+                } else {
+                    return setBlockType(blockType, attrs)(state, dispatch);
+                }
+            }
+        }
+
+        return false;
+    };
+};
```

Comment:
Oh, yeah, I didn't want to sidetrack the current cycle into a massive refactoring effort, but once we have achieved MVP, I think we will need to go over the whole text-editor component and refactor a bunch of parts, just to break stuff out into named functions, because following stuff like this is not easy, and a few months down the line, when there's no one who has it fresh in their memory that you can ask, it's going to be a nightmare.

(Not just a comment about this code specifically, there seems to be a lot of deeply nestled code since earlier too.)

Reply:
I definitely agree!